src/conduits/RwaInputConduit3.sol

```diff
@@ -0,0 +1,219 @@
+
+    modifier auth() {
+        require(wards[msg.sender] == 1, "RwaInputConduit3/not-authorized");
+        _;
+    }
+
+    /**
+     * @notice Define addresses and gives `msg.sender` admin access.
+     * @param _psm PSM contract address.
+     * @param _to RwaUrn contract address.
+     */
+    constructor(address _psm, address _to, address _quitAddress) public {
+        DSTokenAbstract _gem = DSTokenAbstract(GemJoinAbstract(PsmAbstract(_psm).gemJoin()).gem());
+        psm = PsmAbstract(_psm);
+        dai = DSTokenAbstract(PsmAbstract(_psm).dai());
+        gem = _gem;
+        to = _to;
+        quitAddress = _quitAddress;
+
+        // Give unlimited approve to PSM gemjoin
+        _gem.approve(address(PsmAbstract(_psm).gemJoin()), 2**256 - 1);
+
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    /*//////////////////////////////////
+               Authorization
+    //////////////////////////////////*/
+
+    /**
+     * @notice Grants `usr` admin access to this contract.
+     * @param usr The user address.
+     */
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+        emit Rely(usr);
+    }
+
+    /**
+     * @notice Revokes `usr` admin access from this contract.
+     * @param usr The user address.
+     */
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+
+    /**
+     * @notice Grants `usr` push access to this contract.
+     * @param usr The user address.
+     */
+    function mate(address usr) external auth {
+        may[usr] = 1;
+        emit Mate(usr);
+    }
+
+    /**
+     * @notice Revokes `usr` push access from this contract.
+     * @param usr The user address.
+     */
+    function hate(address usr) external auth {
+        may[usr] = 0;
+        emit Hate(usr);
+    }
+
+    /*//////////////////////////////////
+               Administration
+    //////////////////////////////////*/
+
+    /**
+     * @notice Updates a contract parameter.
+     * @param what The changed parameter name. `"quitAddress"`
+     * @param data The new value of the parameter.
+     */
+    function file(bytes32 what, address data) external auth {
+        if (what == "quitAddress") {
+            require(data != address(0), "RwaInputConduit3/invalid-quit-address");
+            quitAddress = data;
+        } else {
+            revert("RwaInputConduit3/unrecognised-param");
+        }
+
+        emit File(what, data);
+    }
+
+    /**
+     * @notice Internal method which:
+     *  - Approve PSM gemJoin Adapter with WAD amount of GEM
+     *  - Swap GEM to DAI through PSM
+     *  - Push DAI to Recipient address (to)
+     * @param gemAmount GEM Amount to swap
+     */
+    function _swapAndPush(uint256 gemAmount) internal {
+        psm.sellGem(address(this), gemAmount);
+
+        uint256 balance = dai.balanceOf(address(this));
+        dai.transfer(to, balance);
+
+        emit Push(to, balance);
+    }
```

Comment:
Maybe now inline this method in `push`, since it's called only once.

Reply:
Done